File: src/main/java/org/example/kalkulationsprogramm/domain/ProjektGeschaeftsdokument.java
package org.example.kalkulationsprogramm.domain;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import lombok.Getter;
import lombok.Setter;

@Entity
@Getter
@Setter

public class ProjektGeschaeftsdokument extends ProjektDokument
{
    @Column(nullable = false)
    private String dokumentid;
    @Column(nullable = false)
    private String geschaeftsdokumentart;
    private LocalDate rechnungsdatum;
    private LocalDate faelligkeitsdatum;
    private BigDecimal bruttoBetrag;
    @Column(nullable = false)
    private boolean bezahlt = false;
    @Enumerated(EnumType.STRING)
    private Mahnstufe mahnstufe;
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "referenz_dokument_id")
    private ProjektGeschaeftsdokument referenzDokument;
    @OneToMany(mappedBy = "referenzDokument", fetch = FetchType.LAZY)
    private List<ProjektGeschaeftsdokument> mahnungen = new ArrayList<>();

    /**
     * {@code true} wenn dieser Eintrag automatisch vom System via
     * {@code AusgangsGeschaeftsDokumentService#erstelleOffenenPostenEintrag} erzeugt wurde.
     * Nur systemgenerierte Rechnungen erhalten automatische Zahlungserinnerungen per E-Mail.
     * Manuell im Offene-Posten-Editor erfasste Einträge haben dieses Flag {@code false}.
     */
    @Column(name = "system_generiert", nullable = false)
    private boolean systemGeneriert = false;

}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/ProjektGeschaeftsdokument.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import java.math.BigDecimal
import java.time.LocalDate

@Entity
open class ProjektGeschaeftsdokument : ProjektDokument() {
    @Column(nullable = false)
    open var dokumentid: String? = null

    @Column(nullable = false)
    open var geschaeftsdokumentart: String? = null

    open var rechnungsdatum: LocalDate? = null

    open var faelligkeitsdatum: LocalDate? = null

    open var bruttoBetrag: BigDecimal? = null

    @Column(nullable = false)
    open var bezahlt: Boolean = false

    @Enumerated(EnumType.STRING)
    open var mahnstufe: Mahnstufe? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "referenz_dokument_id")
    open var referenzDokument: ProjektGeschaeftsdokument? = null

    @OneToMany(mappedBy = "referenzDokument", fetch = FetchType.LAZY)
    open var mahnungen: MutableList<ProjektGeschaeftsdokument> = mutableListOf()

    @Column(name = "system_generiert", nullable = false)
    open var systemGeneriert: Boolean = false

    open fun isBezahlt(): Boolean = bezahlt

    open fun isSystemGeneriert(): Boolean = systemGeneriert

}
